cdc/processor/sinkmanager/table_sink_wrapper.go

```diff
@@ -0,0 +1,156 @@
+	"github.com/pingcap/tiflow/cdc/processor/pipeline"
+	"github.com/pingcap/tiflow/cdc/processor/tablepb"
+	sinkv2 "github.com/pingcap/tiflow/cdc/sinkv2/tablesink"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// tableSinkWrapper is a wrapper of TableSink, it is used in SinkManager to manage TableSink.
+// Because in the SinkManager, we write data to TableSink and RedoManager concurrently,
+// so current sink node can not be reused.
+type tableSinkWrapper struct {
+	// changefeed used for logging.
+	changefeed model.ChangeFeedID
+	// tableID used for logging.
+	tableID model.TableID
+	// tableSink is the underlying sink.
+	tableSink sinkv2.TableSink
+	// state used to control the lifecycle of the table.
+	state *tablepb.TableState
+	// targetTs is the upper bound of the table sink.
+	targetTs model.Ts
+	// receivedSorterResolvedTs is the resolved ts received from the sorter.
+	// We use this to advance the redo log.
+	receivedSorterResolvedTs atomic.Uint64
+}
+
+func newTableSinkWrapper(
+	changefeed model.ChangeFeedID,
+	tableID model.TableID,
+	tableSink sinkv2.TableSink,
+	state *tablepb.TableState,
+	targetTs model.Ts,
+) *tableSinkWrapper {
+	return &tableSinkWrapper{
+		changefeed: changefeed,
+		tableID:    tableID,
+		tableSink:  tableSink,
+		state:      state,
+		targetTs:   targetTs,
+	}
+}
+
+func (t *tableSinkWrapper) appendRowChangedEvents(events ...*model.RowChangedEvent) {
+	t.tableSink.AppendRowChangedEvents(events...)
+}
+
+func (t *tableSinkWrapper) updateReceivedSorterResolvedTs(ts model.Ts) {
+	t.receivedSorterResolvedTs.Store(ts)
+}
+
+func (t *tableSinkWrapper) updateResolvedTs(ts model.ResolvedTs) error {
+	if err := t.tableSink.UpdateResolvedTs(ts); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (t *tableSinkWrapper) getCheckpointTs() model.ResolvedTs {
+	return t.tableSink.GetCheckpointTs()
+}
+
+func (t *tableSinkWrapper) getReceivedSorterResolvedTs() model.Ts {
+	return t.receivedSorterResolvedTs.Load()
+}
+
+func (t *tableSinkWrapper) getState() tablepb.TableState {
+	return t.state.Load()
+}
+
+func (t *tableSinkWrapper) close(ctx context.Context) error {
+	t.state.Store(tablepb.TableStateStopping)
+	// table stopped state must be set after underlying sink is closed
+	defer t.state.Store(tablepb.TableStateStopped)
+	err := t.tableSink.Close(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	log.Info("sink is closed",
+		zap.Int64("tableID", t.tableID),
+		zap.String("namespace", t.changefeed.Namespace),
+		zap.String("changefeed", t.changefeed.ID))
+	return cerror.ErrTableProcessorStoppedSafely.GenWithStackByArgs()
+}
+
+// convertRowChangedEvents uses to convert RowChangedEvents to TableSinkRowChangedEvents.
+// It will deal with the old value compatibility.
+func convertRowChangedEvents(changefeed model.ChangeFeedID, tableID model.TableID, enableOldValue bool, events ...*model.PolymorphicEvent) ([]*model.RowChangedEvent, error) {
```

Comment:
is it reasonable to put this method in this struct?

Reply:
No, it will also be used to redo tasks. It is more like a util.